crates/red_knot_python_semantic/src/semantic_index/visibility_constraints.rs

```diff
@@ -553,37 +554,99 @@ impl VisibilityConstraints {
         }
     }
 
+    fn analyze_single_pattern_predicate_kind<'db>(
+        db: &'db dyn Db,
+        predicate_kind: &PatternPredicateKind<'db>,
+        subject: Expression<'db>,
+    ) -> Truthiness {
+        match predicate_kind {
+            PatternPredicateKind::Value(value) => {
+                let subject_ty = infer_expression_type(db, subject);
+                let value_ty = infer_expression_type(db, *value);
+
+                if subject_ty.is_single_valued(db) {
+                    Truthiness::from(subject_ty.is_equivalent_to(db, value_ty))
+                } else {
+                    Truthiness::Ambiguous
+                }
+            }
+            PatternPredicateKind::Singleton(singleton) => {
+                let subject_ty = infer_expression_type(db, subject);
+
+                let singleton_ty = match singleton {
+                    ruff_python_ast::Singleton::None => Type::none(db),
+                    ruff_python_ast::Singleton::True => Type::BooleanLiteral(true),
+                    ruff_python_ast::Singleton::False => Type::BooleanLiteral(false),
+                };
+
+                if subject_ty.is_single_valued(db) {
+                    Truthiness::from(subject_ty.is_equivalent_to(db, singleton_ty))
+                } else if subject_ty.is_disjoint_from(db, singleton_ty) {
+                    Truthiness::AlwaysFalse
+                } else {
+                    Truthiness::Ambiguous
+                }
+            }
+            PatternPredicateKind::Or(predicates) => {
+                use std::ops::ControlFlow;
+                let (ControlFlow::Break(truthiness) | ControlFlow::Continue(truthiness)) =
+                    predicates
+                        .iter()
+                        .map(|p| Self::analyze_single_pattern_predicate_kind(db, p, subject))
+                        // this is just a "max", but with a slight optimization: `AlwaysTrue` is the "greatest" possible element, so we short-circuit if we get there
+                        .try_fold(Truthiness::AlwaysFalse, |acc, next| match (acc, next) {
+                            (Truthiness::AlwaysTrue, _) | (_, Truthiness::AlwaysTrue) => {
+                                ControlFlow::Break(Truthiness::AlwaysTrue)
+                            }
+                            (Truthiness::Ambiguous, _) | (_, Truthiness::Ambiguous) => {
+                                ControlFlow::Continue(Truthiness::Ambiguous)
+                            }
+                            (Truthiness::AlwaysFalse, Truthiness::AlwaysFalse) => {
+                                ControlFlow::Continue(Truthiness::AlwaysFalse)
+                            }
+                        });
+                truthiness
```

Comment:
This is awesome to have, and a really nice implementation of it too!

I don't see any tests exercising it, though. If I replace this entire arm with just `Truthiness::Ambiguous`, all tests pass. Can we add some tests for this?